pkg/runtime/wfengine/wfengine.go

```diff
@@ -12,199 +12,149 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 	}
 
-	if wfe.executor == nil {
-		return errors.New("gRPC executor is not yet configured")
-	}
-
-	// Register actor backend if backend is actor
-	abe, ok := wfe.Backend.(*actorsbe.ActorBackend)
-	if ok {
-		abe.WaitForActorsReady(ctx)
-		abe.RegisterActor(ctx)
-	}
+	var activeConns uint64
+	var lock sync.Mutex
+	executor, registerGrpcServerFn := backend.NewGrpcExecutor(abackend, log,
+		backend.WithOnGetWorkItemsConnectionCallback(func(ctx context.Context) error {
+			lock.Lock()
+			defer lock.Unlock()
+
+			activeConns++
+			if activeConns == 1 {
+				log.Debug("Registering workflow actors")
+				return abackend.RegisterActors(ctx)
+			}
+
+			return nil
+		}),
+		backend.WithOnGetWorkItemsDisconnectCallback(func(ctx context.Context) error {
+			lock.Lock()
+			defer lock.Unlock()
+
+			activeConns--
+			if activeConns == 0 {
+				log.Debug("Unregistering workflow actors")
+				return abackend.UnRegisterActors(ctx)
+			}
+
+			return nil
+		}),
+	)
 
 	// There are separate "workers" for executing orchestrations (workflows) and activities
 	orchestrationWorker := backend.NewOrchestrationWorker(
-		wfe.Backend,
-		wfe.executor,
+		abackend,
+		executor,
 		wfBackendLogger,
-		backend.WithMaxParallelism(wfe.spec.GetMaxConcurrentWorkflowInvocations()))
+		backend.WithMaxParallelism(opts.Spec.GetMaxConcurrentWorkflowInvocations()))
 	activityWorker := backend.NewActivityTaskWorker(
-		wfe.Backend,
-		wfe.executor,
+		abackend,
+		executor,
 		wfBackendLogger,
-		backend.WithMaxParallelism(wfe.spec.GetMaxConcurrentActivityInvocations()))
-	wfe.worker = backend.NewTaskHubWorker(wfe.Backend, orchestrationWorker, activityWorker, wfBackendLogger)
+		backend.WithMaxParallelism(opts.Spec.GetMaxConcurrentActivityInvocations()))
+	worker := backend.NewTaskHubWorker(abackend, orchestrationWorker, activityWorker, wfBackendLogger)
+
+	return &engine{
+		appID:                opts.AppID,
+		namespace:            opts.Namespace,
+		actors:               opts.Actors,
+		worker:               worker,
+		backend:              abackend,
+		registerGrpcServerFn: registerGrpcServerFn,
+		client: &client{
+			logger: wfBackendLogger,
+			client: backend.NewTaskHubClient(abackend),
+		},
+	}, nil
+}
 
+func (wfe *engine) RegisterGrpcServer(server *grpc.Server) {
+	wfe.registerGrpcServerFn(server)
+}
+
+func (wfe *engine) Run(ctx context.Context) error {
 	// Start the Durable Task worker, which will allow workflows to be scheduled and execute.
 	if err := wfe.worker.Start(ctx); err != nil {
 		return fmt.Errorf("failed to start workflow engine: %w", err)
 	}
 
-	wfe.IsRunning = true
-	wfLogger.Info("Workflow engine started")
-
-	wfe.SetWorkflowEngineReadyDone()
-
-	return nil
-}
-
-func (wfe *WorkflowEngine) Close(ctx context.Context) error {
-	wfe.startMutex.Lock()
-	defer wfe.startMutex.Unlock()
+	log.Info("Workflow engine started")
+	<-ctx.Done()
 
-	if !wfe.IsRunning {
-		return nil
+	if err := wfe.worker.Shutdown(context.Background()); err != nil {
```

Comment:
Ik this is after the `ctx.Done()`, but should this be wrapped in a defer func to guarantee the shutdown?

Reply:
Shutdown is always guaranteed if Start returned successfully.